test/protocol/eval_test.rb

```diff
@@ -57,4 +57,57 @@ def test_eval_evaluates_arithmetic_expressions
       end
     end
   end
+
+  class EvaluateThreadTest < ProtocolTestCase
+    def test_eval_doesnt_deadlock
+      program = <<~RUBY
+        1| th0 = Thread.new{sleep}
+        2| m = Mutex.new; q = Queue.new
+        3| th1 = Thread.new do
+        4|   m.lock; q << true
+        5|   sleep 1
+        6|   m.unlock
+        7| end
+        8| q.pop # wait for locking
+        9| p :ok
+      RUBY
+
+      run_protocol_scenario program, cdp: false do
+        req_add_breakpoint 9
+        req_continue
+        assert_repl_result({value: 'false', type: 'FalseClass'}, 'm.lock.nil?', frame_idx: 0)
+        req_continue
+      end
+    end
+
+    def test_eval_stops_threads_after_finished
+      program = <<~RUBY
+        1| count = 0
+        2| th0 = Thread.new do
+        3|   loop do
+        4|     count += 1
+        5|     sleep 0.05
+        6|   end
+        7| end
+        8|
+        9| sleep 0.1
+       10|
+       11| __LINE__
+      RUBY
+
+      run_protocol_scenario program, cdp: false do
+        req_add_breakpoint 11
+        req_continue
+        assert_repl_result({value: '1', type: 'Integer'}, '1', frame_idx: 0)
+        locals = gather_variables
+        count_var_value_1 = locals.find{|v| v[:name] == 'count'}[:value]
+        locals = gather_variables
+        count_var_value_2 = locals.find{|v| v[:name] == 'count'}[:value]
+        # if the thread is stopped, the value of count will not be changed.
+        assert_equal(count_var_value_1, count_var_value_2)
```

Comment:
This test is not acceptable because `sleep 0.05` introducing nondeterministic results.

Reply:
I've updated the test to not use `sleep`. And the test failure doesn't seem to be related (another [occurrence](https://github.com/ruby/debug/actions/runs/4516497233/jobs/7954891876))